src/main/scala/is/hail/methods/PCA.scala

```diff
@@ -56,12 +57,66 @@ trait PCA {
     (svd.V.multiply(DenseMatrix.diag(svd.s)), optionLoadings, someIf(computeEigenvalues, svd.s.toArray.map(math.pow(_, 2))))
   }
 
-  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Variant]], IndexedRowMatrix)
+  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Any]], IndexedRowMatrix)
+}
+
+class ExprPCA(val expr: String) extends PCA {
+  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Any]], IndexedRowMatrix) = {
+
+    val partitionSizes = vsm.rdd2.mapPartitions(it => Iterator.single(it.size)).collect().scanLeft(0)(_ + _)
+    assert(partitionSizes.length == vsm.rdd2.getNumPartitions + 1)
+    val pSizeBc = vsm.sparkContext.broadcast(partitionSizes)
+
+    val rowType = vsm.rowType
+    val samplesBc = vsm.sampleIdsBc
+    val sampleAnnotationsBc = vsm.sampleAnnotationsBc
+
+    val ec = EvalContext(Map(
+      "global" -> (0, vsm.globalSignature),
+      "v" -> (1, vsm.vSignature),
+      "va" -> (2, vsm.vaSignature),
+      "s" -> (3, vsm.sSignature),
+      "sa" -> (4, vsm.saSignature),
+      "g" -> (5, vsm.genotypeSignature)))
+    val f = RegressionUtils.parseExprAsDouble(expr, ec)
+    ec.set(0, vsm.globalAnnotation)
+
+    val mat = vsm.rdd2.mapPartitionsWithIndex {case (i, it) =>
+      val pStartIdx = pSizeBc.value(i)
+      var j = 0
+      val ur = new UnsafeRow(rowType)
+      it.map {rv =>
+        ur.set(rv)
+        ec.set(1, ur.get(1))
+        ec.set(2, ur.get(2))
+        val gs = ur.getAs[IndexedSeq[Any]](3)
+        val row = IndexedRow(pStartIdx + j, Vectors.dense(gs.indices.map {k =>
```

Comment:
still a vector!
```scala
scala> val a = Array(1,2,3,4,5)
a: Array[Int] = Array(1, 2, 3, 4, 5)

scala> a.indices.map(x => x * 2)
res0: scala.collection.immutable.IndexedSeq[Int] = Vector(0, 2, 4, 6, 8)
```

`Array.tabulate` won't allocate any intermediates. The fastest possible thing is probably
```scala
val ns = gs.length()
val a = new Array[Double](ns)
var k = 0
while (k < ns) {
  ec.set(3, samplesBc.value(k))
  ec.set(4, sampleAnnotationsBc.value(k))
  ec.set(5, gs(k))
  a(k) = f().toDouble
  k += 1
}
val row = IndexedRow(pStartIdx + j, a)
```
The two are probably comparable after JIT optimization, probably.